openedx_learning/apps/authoring/containers/api.py

```diff
@@ -0,0 +1,380 @@
+            container_id=container_pk,
+            defined_list=entity_list,
+            initial_list=entity_list,
+            # Would frozen_list be always None the 1st time a container is created?
+            frozen_list=None,
+        )
+    return container_version
+
+
+def create_next_container_version(
+    container_pk: int,
+    title: str,
+    publishable_entities_pk: list[int],
+    draft_version_pks: list[int | None],
+    published_version_pks: list[int | None],
+    entity: PublishableEntity,
+    created: datetime,
+    created_by: int | None,
+) -> ContainerEntityVersion:
+    """
+    Create the next version of a container.
+
+    Args:
+        container_pk: The ID of the container to create the next version of.
+        title: The title of the container.
+        publishable_entities_pk: The IDs of the members of the container.
+        entity: The entity that the container belongs to.
+        created: The date and time the container version was created.
+        created_by: The ID of the user who created the container version.
+
+    Returns:
+        The newly created container version.
+    """
+    with atomic():
+        container = ContainerEntity.objects.get(pk=container_pk)
+        last_version = container.versioning.latest
+        next_version_num = last_version.version_num + 1
+        publishable_entity_version = publishing_api.create_publishable_entity_version(
+            entity.pk,
+            version_num=next_version_num,
+            title=title,
+            created=created,
+            created_by=created_by,
+        )
+        # 1. The changes provoking a new version are always the addition, removal of members or reordering. How can we detect changes only in the metadata?
+        # 2. Published and draft versions are always the latest for all members.
+        # 3. When creating a new version, a new user-defined entity list is created to preserve the latest state as the previous user-defined list.
+        # 4. When creating a new version, a new frozen entity list is created copying the state of the user-defined list of the previous version.
+        # 5. While copying the versions into the new frozen list, versions are pinned in new rows for published/draft versions.
+        new_user_defined_list = create_entity_list_with_rows(
+            entity_pks=publishable_entities_pk,
+            draft_version_pks=draft_version_pks,
+            published_version_pks=published_version_pks,
+        )
+        new_frozen_list = copy_rows_to_new_entity_list(
+            rows=get_defined_list_rows_for_container_version(last_version)
+        )
+
+        container_version = ContainerEntityVersion.objects.create(
+            publishable_entity_version=publishable_entity_version,
+            container_id=container_pk,
+            defined_list=new_user_defined_list,
+            initial_list=last_version.initial_list,
+            frozen_list=new_frozen_list,
+        )
```

Comment:
Here, I'm creating a new defined entity list (with new rows) for each new version and also copying things over from the previously defined list into a new frozen list (also, with new rows). This is done to comply with:

1. defined_list should never change for a given ContainerEntityVersion to simplify reasoning.
2. The defined_list is the EntityList that the author defines. This should never change, even if the things it references get soft-deleted (because we'll need to maintain it for reverts).

But we also have this requirement:

3. Ordering should be treated as immutable–if the ordering needs to change, we create a new EntityList and copy things over.

So no new entity list rows for defined lists should be created if the ordering doesn't change or versions are not pinned, which I'm currently not considering. I remember discussing this with Dave a couple of weeks ago too, so implementing it seems like the right approach here.

Reply:
I'm testing my changes out before pushing, I'm hoping I can do it today.